### PR TITLE
Fix writing the fields in GRIB to files

### PR DESCRIPTION
**Pull request by ninaburg**
- processing.py:  Fix writing grib1 and grib2 fields to file by removing the condition on the attribute that didn't exist and thus nothing was wirtten in the file. 
- flexpart_utils.py: Reassign the attributes to the dataarray which were lost when doing: `ds_out["cp"] = ds_out["cp"] * 1000`

### CODE REVIEW

File: flexprep/domain/processing.py [hunk 1]
```diff
@@ -122,20 +121,40 @@ def _save_output(
...
+                        else:
+                            msg = metadata.override(
+                                ref.message, shortName=field.parameter["shortName"]
+                            )
+                        field.attrs = msg
+                    grib_decoder.save(field, output_file)
+            logger.info("Writing GRIB fields to file completed.")
+
+            # Upload the file to S3
+            S3client().upload_file(output_file.name, key=key)
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]: Why did you move the upload out of the context manager ?

File: flexprep/domain/flexpart_utils.py [hunk 2]
```diff
@@ -42,13 +42,19 @@ def prepare_output(
...
+        ds_out[field] = ds_in[field].isel(lead_time=-1)
     for field in missing_const:
-        ds_out[field] = ds_in[field].squeeze()
+        ds_out[field] = ds_in[field]
 
     ds_out["etadot"] = ds_out.pop("omega")
+
+    cp_attrs = ds_out["cp"].attrs.copy()
     ds_out["cp"] = ds_out["cp"] * 1000
+    ds_out["cp"] = ds_out["cp"].assign_attrs(cp_attrs)
```
Comment (cfkanesan) [inline flexprep/domain/flexpart_utils.py]:
```suggestion
    ds_out["cp"] = (ds_out["cp"] * 1000).assign_attrs(ds_out["cp"].attrs)
```

File: flexprep/domain/processing.py [hunk 3]
```diff
@@ -122,20 +121,40 @@ def _save_output(
         forecast_ref_time_str = forecast_ref_time.strftime("%Y%m%d%H%M")
 
         try:
-            key = f"output_dispf{forecast_ref_time_str}_step{step_to_process}"
+            key = f"output_dispf{forecast_ref_time_str}{step_to_process}"
+
+            ref = next(
+                f
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]: I know I wrote this but maybe it's better to spell out `field`

File: flexprep/domain/processing.py [hunk 4]
```diff
@@ -122,20 +121,40 @@ def _save_output(
...
-                with open(output_file.name, "wb") as fout:
-                    for name, field in ds_out.items():
-                        if field.attrs.get("v_coord") == "hybrid":
-                            logger.info(f"Writing GRIB fields to {output_file.name}")
-                            grib_decoder.save(field, fout)
-
-                    # Upload the file to S3
-                    S3client().upload_file(output_file.name, key=key)
+                for name, field in ds_out.items():
+                    if field.squeeze().isnull().all():
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]: Why squeeze?

**Review by cfkanesan — COMMENTED** [review]
I have some questions about the changes

File: flexprep/domain/processing.py (hunk 4 above)
Comment (ninaburg) [reply flexprep/domain/processing.py]: I wanted to be sure that the check is not done on axis separately but it this case the .all() takes care to reduce it to one bool. So yes .squeeze() is not needed.

File: flexprep/domain/processing.py (hunk 1 above)
Comment (ninaburg) [reply flexprep/domain/processing.py]: I moved it outside the context manager to ensure the file is fully written and closed before proceeding with the upload. Does it make more sense for you to leave it inside?

Comment (cfkanesan) [reply flexprep/domain/processing.py]: File operations are blocking by default, meaning that the file contents are updated (no longer in the buffer) when the call returns. The purpose of the context manager is to remove the file if an exception is encountered. In this case, it makes sense to me to leave the upload inside the context manager.

File: flexprep/domain/processing.py [hunk 5]
```diff
@@ -122,20 +121,48 @@ def _save_output(
...
+
+            ref = next(
+                field
+                for field in ds_out.values()
+                if metadata.extract_keys(field.message, "editionNumber") == 2
+                and metadata.extract_keys(
+                    field.message, "productDefinitionTemplateNumber"
+                )
+                == 0
+            )
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]:
```suggestion
            ref_keys = "editionNumber", "productDefinitionTemplateNumber"
            ref_values = 2, 0
            ref = next(
                field
                for field in ds_out.values()
                if metadata.extract_keys(field.message, ref_keys) == ref_values
            )
```

**Review by cfkanesan — APPROVED** [review]
LGTM with one minor suggestion
